src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,153 @@
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index is out of bounds");
+        }
+        if (index == size) {
+            Node<T> previousTail = tail;
+            tail = new Node<>();
+            tail.previous = previousTail;
+            tail.value = value;
+            if (previousTail == null) {
+                head = tail;
+            } else {
+                previousTail.next = tail;
+            }
+        } else {
+            Node<T> nodeBeforeInput = getNode(index);
+            Node<T> oldPrevious = nodeBeforeInput.previous;
+            Node<T> inputNode = new Node<>();
+            inputNode.previous = oldPrevious;
+            inputNode.value = value;
+            inputNode.next = nodeBeforeInput;
+            nodeBeforeInput.previous = inputNode;
+            if (oldPrevious == null) {
+                head = inputNode;
+            } else {
+                oldPrevious.next = inputNode;
+            }
+        }
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T value : list) {
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNode(index);
+        T previousValue = node.value;
+        node.value = value;
+        return previousValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> target = getNode(index);
+        removingAction(target);
+        return target.value;
     }
 
     @Override
     public boolean remove(T t) {
+        Node<T> node = head;
+        while (node != null) {
+            if (node.value == t || node.value.equals(t)) {
+                removingAction(node);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index is out of bounds");
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndex(index);
+        Node<T> node = head;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        return node;
+    }
+
+    private void removingAction(Node<T> target) {
+        Node<T> previous = target.previous;
+        Node<T> next = target.next;
+
+        if (previous == null) {
+            head = next;
+        } else {
+            previous.next = next;
```

Comment:
```suggestion
            previous.next = next;
            target.previous = null;
```
in this task it is important to understand the concept of links
so let`s remove the links from both sides

Reply:
got it, added